File: ios/Hadithly/Core/Purchases/PurchaseManager.swift
import Foundation
import Observation
import RevenueCat

/// RevenueCat lifecycle and purchase state. The SDK is configured once at
/// launch, then associated with the verified Clerk user after sign-in.
@MainActor
@Observable
final class PurchaseManager {
    private(set) var packages: [Package] = []
    private(set) var isConfigured = false
    private(set) var isLoading = false
    private(set) var isPurchasing = false
    private(set) var hasProEntitlement = false
    private(set) var errorMessage: String?

    init(apiKey: String) {
        let trimmed = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard
            !trimmed.isEmpty,
            !trimmed.localizedCaseInsensitiveContains("placeholder"),
            !trimmed.localizedCaseInsensitiveContains("your_"),
            !trimmed.lowercased().hasPrefix("test_")
        else { return }

        #if DEBUG
        Purchases.logLevel = .debug
        #else
        Purchases.logLevel = .warn
        #endif
        Purchases.configure(withAPIKey: trimmed)
        isConfigured = true
    }

    func logIn(appUserID: String) async {
        guard isConfigured else { return }
        let result: (CustomerInfo?, Error?) = await withCheckedContinuation { continuation in
            Purchases.shared.logIn(appUserID) { customerInfo, _, error in
                continuation.resume(returning: (customerInfo, error))
            }
        }
        apply(customerInfo: result.0)
        if let error = result.1 { errorMessage = error.localizedDescription }
    }

    func logOut() async {
        guard isConfigured, !Purchases.shared.isAnonymous else { return }
        let result: (CustomerInfo?, Error?) = await withCheckedContinuation { continuation in
            Purchases.shared.logOut { customerInfo, error in
                continuation.resume(returning: (customerInfo, error))
            }
        }
        apply(customerInfo: result.0)
        if let error = result.1 { errorMessage = error.localizedDescription }
    }

    func loadOffering() async {
        guard isConfigured else {
            errorMessage = "Subscriptions are not configured for this build."
            return
        }
        isLoading = true
        errorMessage = nil
        let result: (Offerings?, Error?) = await withCheckedContinuation { continuation in
            Purchases.shared.getOfferings { offerings, error in
                continuation.resume(returning: (offerings, error))
            }
        }
        isLoading = false
        packages = Self.sortedPackages(result.0?.current?.availablePackages ?? [])
        if packages.isEmpty, result.1 == nil {
            errorMessage = "No subscription plans are available right now."
        } else if let error = result.1 {
            errorMessage = error.localizedDescription
        }
    }

    func purchase(_ package: Package) async -> Bool {
        guard isConfigured else { return false }
        isPurchasing = true
        errorMessage = nil
        let result: (CustomerInfo?, Error?, Bool) = await withCheckedContinuation { continuation in
            Purchases.shared.purchase(package: package) { _, customerInfo, error, userCancelled in
                continuation.resume(returning: (customerInfo, error, userCancelled))
            }
        }
        isPurchasing = false
        apply(customerInfo: result.0)
        if let error = result.1, !result.2 { errorMessage = error.localizedDescription }
        return hasProEntitlement
    }

    func restore() async -> Bool {
        guard isConfigured else { return false }
        isPurchasing = true
        errorMessage = nil
        let result: (CustomerInfo?, Error?) = await withCheckedContinuation { continuation in
            Purchases.shared.restorePurchases { customerInfo, error in
                continuation.resume(returning: (customerInfo, error))
            }
        }
        isPurchasing = false
        apply(customerInfo: result.0)
        if let error = result.1 { errorMessage = error.localizedDescription }
        return hasProEntitlement
    }

    private func apply(customerInfo: CustomerInfo?) {
        guard let customerInfo else { return }
        hasProEntitlement = customerInfo.entitlements["pro"]?.isActive == true
    }

    private static func sortedPackages(_ packages: [Package]) -> [Package] {
        let priority: [PackageType: Int] = [.annual: 0, .sixMonth: 1, .threeMonth: 2, .twoMonth: 3, .monthly: 4, .weekly: 5]
        return packages.sorted {
            (priority[$0.packageType] ?? 99) < (priority[$1.packageType] ?? 99)
        }
    }
}

File: android/app/src/main/kotlin/com/hadithly/app/core/purchases/PurchaseManager.kt
package com.hadithly.app.core.purchases

import android.app.Activity
import android.content.Context
import com.hadithly.app.BuildConfig
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.LogLevel
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PackageType
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.getOfferingsWith
import com.revenuecat.purchases.logInWith
import com.revenuecat.purchases.logOutWith
import com.revenuecat.purchases.purchaseWith
import com.revenuecat.purchases.restorePurchasesWith
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update

/** RevenueCat lifecycle and paywall state, keyed to the verified Clerk user. */
class PurchaseManager(context: Context, apiKey: String) {

    data class State(
        val packages: List<Package> = emptyList(),
        val configured: Boolean = false,
        val loading: Boolean = false,
        val purchasing: Boolean = false,
        val hasProEntitlement: Boolean = false,
        val errorMessage: String? = null,
    )

    private val _state = MutableStateFlow(State())
    val state: StateFlow<State> = _state

    init {
        val trimmed = apiKey.trim()
        if (
            trimmed.isNotEmpty() &&
            !trimmed.contains("placeholder", ignoreCase = true) &&
            !trimmed.contains("your_", ignoreCase = true)
        ) {
            Purchases.logLevel = if (BuildConfig.DEBUG) LogLevel.DEBUG else LogLevel.WARN
            Purchases.configure(PurchasesConfiguration.Builder(context, trimmed).build())
            _state.update { it.copy(configured = true) }
        }
    }

    fun logIn(appUserId: String) {
        if (!_state.value.configured) return
        Purchases.sharedInstance.logInWith(
            appUserId,
            onError = { error -> _state.update { it.copy(errorMessage = error.message) } },
            onSuccess = { customerInfo, _ -> apply(customerInfo) },
        )
    }

    fun logOut() {
        if (!_state.value.configured || Purchases.sharedInstance.isAnonymous) return
        Purchases.sharedInstance.logOutWith(
            onError = { error -> _state.update { it.copy(errorMessage = error.message) } },
            onSuccess = ::apply,
        )
    }

    fun loadOffering() {
        if (!_state.value.configured) {
            _state.update { it.copy(errorMessage = "Subscriptions are not configured for this build.") }
            return
        }
        _state.update { it.copy(loading = true, errorMessage = null) }
        Purchases.sharedInstance.getOfferingsWith(
            onError = { error -> _state.update { it.copy(loading = false, errorMessage = error.message) } },
            onSuccess = { offerings ->
                val packages = offerings.current?.availablePackages.orEmpty().sortedBy(::packagePriority)
                _state.update {
                    it.copy(
                        packages = packages,
                        loading = false,
                        errorMessage = if (packages.isEmpty()) "No subscription plans are available right now." else null,
                    )
                }
            },
        )
    }

    fun purchase(activity: Activity, packageToPurchase: Package, onUnlocked: () -> Unit) {
        if (!_state.value.configured) return
        _state.update { it.copy(purchasing = true, errorMessage = null) }
        Purchases.sharedInstance.purchaseWith(
            PurchaseParams.Builder(activity, packageToPurchase).build(),
            onError = { error, userCancelled ->
                _state.update {
                    it.copy(purchasing = false, errorMessage = if (userCancelled) null else error.message)
                }
            },
            onSuccess = { _, customerInfo ->
                apply(customerInfo)
                _state.update { it.copy(purchasing = false) }
                if (_state.value.hasProEntitlement) onUnlocked()
            },
        )
    }

    fun restore(onUnlocked: () -> Unit) {
        if (!_state.value.configured) return
        _state.update { it.copy(purchasing = true, errorMessage = null) }
        Purchases.sharedInstance.restorePurchasesWith(
            onError = { error -> _state.update { it.copy(purchasing = false, errorMessage = error.message) } },
            onSuccess = { customerInfo ->
                apply(customerInfo)
                _state.update { it.copy(purchasing = false) }
                if (_state.value.hasProEntitlement) onUnlocked()
            },
        )
    }

    private fun apply(customerInfo: CustomerInfo) {
        _state.update {
            it.copy(hasProEntitlement = customerInfo.entitlements["pro"]?.isActive == true)
        }
    }

    private fun packagePriority(packageToSort: Package): Int = when (packageToSort.packageType) {
        PackageType.ANNUAL -> 0
        PackageType.SIX_MONTH -> 1
        PackageType.THREE_MONTH -> 2
        PackageType.TWO_MONTH -> 3
        PackageType.MONTHLY -> 4
        PackageType.WEEKLY -> 5
        else -> 99
    }
}
